Source/WebCore/style/StyleOrderedNamedLinesCollector.h

```diff
@@ -0,0 +1,175 @@
+    {
+    }
+    virtual ~OrderedNamedLinesCollector() = default;
+
+    bool isEmpty() const { return m_orderedNamedGridLines.map.isEmpty() && m_orderedNamedAutoRepeatGridLines.map.isEmpty(); }
+    virtual void collectLineNamesForIndex(Vector<String>&, unsigned index) const = 0;
+
+    virtual int namedGridLineCount() const { return m_orderedNamedGridLines.map.size(); }
+
+protected:
+    enum class NamedLinesType : bool { NamedLines, AutoRepeatNamedLines };
+    void appendLines(Vector<String>&, unsigned index, NamedLinesType) const;
+
+    const OrderedNamedGridLinesMap& m_orderedNamedGridLines;
+    const OrderedNamedGridLinesMap& m_orderedNamedAutoRepeatGridLines;
+};
+
+class OrderedNamedLinesCollectorInGridLayout : public OrderedNamedLinesCollector {
+public:
+    OrderedNamedLinesCollectorInGridLayout(ExtractorState& state, bool isRowAxis, unsigned autoRepeatTracksCount, unsigned autoRepeatTrackListLength)
+        : OrderedNamedLinesCollector(state, isRowAxis)
+        , m_insertionPoint(isRowAxis ? state.style.gridAutoRepeatColumnsInsertionPoint() : state.style.gridAutoRepeatRowsInsertionPoint())
+        , m_autoRepeatTotalTracks(autoRepeatTracksCount)
+        , m_autoRepeatTrackListLength(autoRepeatTrackListLength)
+    {
+    }
+
+    void collectLineNamesForIndex(Vector<String>&, unsigned index) const override;
+
+private:
+    unsigned m_insertionPoint;
+    unsigned m_autoRepeatTotalTracks;
+    unsigned m_autoRepeatTrackListLength;
+};
+
+class OrderedNamedLinesCollectorInSubgridLayout : public OrderedNamedLinesCollector {
+public:
+    OrderedNamedLinesCollectorInSubgridLayout(ExtractorState& state, bool isRowAxis, unsigned totalTracksCount)
+        : OrderedNamedLinesCollector(state, isRowAxis)
+        , m_insertionPoint(isRowAxis ? state.style.gridAutoRepeatColumnsInsertionPoint() : state.style.gridAutoRepeatRowsInsertionPoint())
+        , m_autoRepeatLineSetListLength((isRowAxis ? state.style.autoRepeatOrderedNamedGridColumnLines() : state.style.autoRepeatOrderedNamedGridRowLines()).map.size())
+        , m_totalLines(totalTracksCount + 1)
+    {
+        if (!m_autoRepeatLineSetListLength) {
+            m_autoRepeatTotalLineSets = 0;
+            return;
+        }
+        unsigned named = (isRowAxis ? state.style.orderedNamedGridColumnLines() : state.style.orderedNamedGridRowLines()).map.size();
+        if (named >= m_totalLines) {
+            m_autoRepeatTotalLineSets = 0;
+            return;
+        }
+        m_autoRepeatTotalLineSets = (m_totalLines - named) / m_autoRepeatLineSetListLength;
+        m_autoRepeatTotalLineSets *= m_autoRepeatLineSetListLength;
+    }
+
+    void collectLineNamesForIndex(Vector<String>&, unsigned index) const override;
+
+    int namedGridLineCount() const override { return m_totalLines; }
+private:
+    unsigned m_insertionPoint;
+    unsigned m_autoRepeatTotalLineSets;
```

Comment:
```suggestion
    unsigned m_autoRepeatTotalLineSets { 0 };
```
Then we could simplify the constructor.